From 26d2ce04c294c53a2043180d7a4dcf35f15bc909 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Fri, 5 Apr 2019 16:51:52 +0200
Subject: [PATCH] clocksource/arm_arch_timer: dovetail: fix call to
 arch-specific init

Due to the definition of arch_clocksource_arch_timer_init missing, the
arch_clocksource_arch_timer_init() handler would not be called on
ARM32, meaning that the vDSO code would not use direct access to the
CP15 counter. Fix this, calling a handler the arch-specific code can
override instead (weak linkage).

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 arch/arm/kernel/vdso.c               | 5 ++++-
 drivers/clocksource/arm_arch_timer.c | 6 ++++--
 2 files changed, 8 insertions(+), 3 deletions(-)

diff --git a/arch/arm/kernel/vdso.c b/arch/arm/kernel/vdso.c
index 0685f3c5833e..6c41cbcc96b0 100644
--- a/arch/arm/kernel/vdso.c
+++ b/arch/arm/kernel/vdso.c
@@ -284,5 +284,8 @@ void arch_clocksource_arch_timer_init(struct clocksource *cs)
 {
 	struct arch_clocksource_data *d = &cs->archdata;
 
-	d->clock_type = ARM_CLOCK_ARCH_TIMER;
+	if (!cs->archdata.vdso_direct)
+		d->clock_type = ARM_CLOCK_NONE;
+	else
+		d->clock_type = ARM_CLOCK_ARCH_TIMER;
 }
diff --git a/drivers/clocksource/arm_arch_timer.c b/drivers/clocksource/arm_arch_timer.c
index 81d312e7f26a..2ea0d1622ace 100644
--- a/drivers/clocksource/arm_arch_timer.c
+++ b/drivers/clocksource/arm_arch_timer.c
@@ -993,6 +993,10 @@ struct arch_timer_kvm_info *arch_timer_get_kvm_info(void)
 	return &arch_timer_kvm_info;
 }
 
+void __weak arch_clocksource_arch_timer_init(struct clocksource *cs)
+{
+}
+
 static void __init arch_counter_register(unsigned type)
 {
 	u64 start_count;
@@ -1007,9 +1011,7 @@ static void __init arch_counter_register(unsigned type)
 
 		clocksource_counter.archdata.vdso_direct = vdso_default;
 
-#ifdef arch_clocksource_arch_timer_init
 		arch_clocksource_arch_timer_init(&clocksource_counter);
-#endif
 	} else {
 		arch_timer_read_counter = arch_counter_get_cntvct_mem;
 	}
-- 
2.16.4

